man/newscacheclean.8.in
.TH newscacheclean 8 "June 2004"
.SH NAME
test \- Purge the NewsCache to SpoolSize (newscache.conf parameter).
.SH SYNOPSIS
.B newscacheclean 
.RI [ \-h ]
.RI [ \-s ] 
.RI [ \-p ]
.RI [ \-\-version ]
.RI [ \-c\ config-file ] 
.br
.SH DESCRIPTION
The 
.B newscacheclean
command purges the Cache from NewsCache to the SpoolSize parameter specified in the newscache.conf file. The program newscacheclean walks through the directory tree of SpoolDirectory (newscache.conf parameter) and adds every newsgroup db and every articel file into the purge list. Every outdated external article is also inserted. Then the purge list is sorted by access time. The purge removes first all outdated external articles and then one article or newsgroup database after the other until SpoolSize (newscache.conf parameter) is reached.

.SH OPTIONS
This program follows the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.
.TP
.B \-h, \-\-help
Show summary of options.
.TP
.B \-v, \-\-version
Show version of program.
.TP
.B \-c \-\-configuration config-file
.TP
.B \-s \-\-statistic
Show statistic informations.
.TP
.B \-p \-\-print-purgetable
Print the sorted purge table.
.TP
.B \-t \-\-try
Try, not really removing files.

.SH FILES
@SYSCONFDIR@/newscache.conf
.br
@SBINDIR@/newscache
.br
@SBINDIR@/newscacheclean
.br
@SBINDIR@/newscacheclean
.br
@SBINDIR@/updatenews


.SH SEE ALSO
.BR newscache (8),
.BR updatenews (8),
.BR newscache.conf (5),

.SH AUTHORS
1997-2000 Thomas Gschwind <tom@infosys.tuwien.ac.at>
.br
2001-2004 Herbert Straub <h.straub@aon.at>

.SH REPORTING BUGS
Report bugs to <newscache@linuxhacker.org>.

.SH COPYRIGHT
.B NewsCache
is released under the
.B GNU General Public License (GPL).
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.



NEWS
Release 1.2:
Herbert Straub - 2004-09-28
New functions:
* Can handle different PAM Servicenames per network, see man
  newscache.conf setcion: AccessSpec Configuration Options -> PAMServicename
  
* Only authenticated clients in a network (new option: authentication) -
  see man newscache.conf section: AccessSpec Configuration Options

* new newcache command option: -o --> print out the configurations options

* new meaning of option: updatenews -a --> force the update of the ActiveDB

* detecting corrupt database files and do a rename of this defect file. In
  such cases NewsCache logs this situation to news.err.

* newscacheclean can also handle corrupt database files. The first run renames
  the bad file and (at the moment) abort further operations. The next run of
  newscacheclean can do work without errors.

New Docus:
* doc/newscache-coredump.txt
  how i catch a coredump

* doc/newscache-debugging:
  how i debug a newscache session

* doc/newscache-pam.txt
  how i use the pam functionality

* doc/newscache-stunnel-ssl.txt
  nntps (SSL) with NewsCache and stunnel

* doc/newscache-corrupt-db-files.txt
  Curropt database files (recory)

Manual Pages:
* all configuration dependig pathes are now configured by the configuration
script.

Bugfixes:
* Compiler Error with -O3 (see Changelog for details)

* internal Command MAP handling

* xdebug dump authorization doesn't display the correct access_flags

* Fixing segfault - Don't read the newscache.conf in the right manner



Release 1.1.92:
* Bugfix Release
  Mike Harding reports a tin hang error, should be fixed with 
  this Release.

Release 1.1.90:
* Completely reworked authentication model: AUTHINFO USER/PASS supports now
	- file (plain or encrypted passwords). configuration option
	  --enable-plainpass (default=no)
	- unix; configuration options:
	  --with-md5 --with-md5auto --with-shadow
	- pam (pluggable authentication modules) confiuration option
	  --with-pam
	- pam+file same as above, but the user permission can be
	  specified in the addiational file
  file and unix authentication methods are also available in earlier
  version, but now all methods are consolidated in a uniform schema.
  All methods understand further group specification and allow 
  additional methods (example: post). Only the commands are
  displayed with the HELP command, for those are the authorization
  is granted.

* authentification errors now handled.
  a client process accepted only three wrong passwords and
  then the client is disconnected. The answer to the client
  is also time delayed. The necessary parameters are now
  fix implemented, but can be configurable in the 
  configuration file.

* extented XDEBUG command
  Now there are two supported commands:
  XDEBUG DUMP AUTHORIZATION
  XDEBUG DUMP CONFIGURATION
  This commands are only available, if the debug option is
  listed in allow parameter.

* a lot of consolidation work is done, to bring the man
  pages, the example configuration file and the source
  code in synchronization. The newscache.conf manpage
  contains only tested and verified parameters (I hope
  so). 

* all programs follows  the  usual GNU command line syntax,
  with long options starting with two dashes (`-').

* all programs options are verified and corrected and(!) a
  description is available in the man pages.
  
* many bugfixes and patches fom Tilman Linneweh (see
  Changelog for details.




Release 1.1.12:

* NewsCacheClean Enhencement: NewsCacheClean missing to purge "outdated"
article files in the spool cache. If the upstream newsserver reduce the
available articles in a newsgroup (ex. from 1-100 to 50-120), then all articles from nr 1-49 and size greater then 15kBytes, are left in the cache (.art* files). Now NewsCacheClean analyzing every Newsgroup DB and mark every outdated .art file in the internal purge table with timestamp 0 and therefore newscacheclean purges all outdated .art files. This new optimized purge method reduce the cache of my NewsCache Application from 4 to 1 GB, means 3 GB outdated articles!

* Catching "caught signal 6" error messages in news.err. This error occours with broken client or upstream server connections

* Require Socket++ 1.12.9

Release 1.1.11:

* doxygen comments in *.h files; configuration available:

* Bugfixes: if configured with --with-syslog=no (see ChangeLog for details)

* Removed: accesslog, cachestat, hoststat, hostlog

* The programs NewsCache and NewsCacheClean are now lowercase

* Bugfix: caught signal 6 (see ChangeLog for details)

* Qualifier -t removed from updatenews and newscacheclean

* A new manual page for newscacheclean 8

* Tested with RedHat Advanced Server (Itanium II) with g++ 3.2


Release 1.1:

The main goal for version 1.1 is porting the source to gcc 3.0 standards.
See documentation ftp://gcc.gnu.org/pub/gcc/libstdc++/libstdc++-doxygen-3.0.tar.gz 

NewsCache uses now the Socket++ Library Version 1.12. Please install this first. See: http://members.aon.at/Socket++/

Release 1.1.6: ready for RedHat 8.0 (tested with socket++ shared library).
RedHat 8.0 use g++ 3.2!

* New Article Spool Management for offline mode (new classes ArtSpooler,
  ObjLock).

* NewsCache, updatenews NewsCacheClean are using own names for syslog
  messages per default. In previous versions all programs are using 
  NewsCache. With ./configure --with-uniquename it is possible, to 
  switch to the old naming standard.



src/NewsCacheClean.cc
/* NewsCacheClean.cc 
 * -- (c) 1997-1998 by Thomas GSCHWIND <tom@infosys.tuwien.ac.at>
 * -- removes the least recently used articles/groups until the 
 * -- size of the spool directory is below SpoolSize
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */
/*
 * Copyright (C) 2001-2003 Herbert Straub
 * - change from chain to table (performance)
 * - adding nice value for NewsCacheClean
 * - removing outdated .art files from resized .db files
 * - adding option -t (try_flag)
 * - adding option -p (print purge table) - debug option
 * 2003-03-11 Herbert Straub
 * - changing table implemention with STL vector
 * - removing all structures and function implementing the table 
 */
#include "config.h"

#include <string.h>
#include <getopt.h>
#include <sys/param.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <sys/types.h>
#include <dirent.h>
#include <unistd.h>
#ifdef TIME_WITH_SYS_TIME
#include <sys/time.h>
#endif
#ifdef HAVE_SETPRIORITY
#include <sys/resource.h>
#endif

#include <string>
#include <vector>
#include <algorithm>

#include "Debug.h"
#include "setugid.h"
#include "Config.h"
#include "Logger.h"
#include "NServer.h"

using namespace std;

#ifndef WITH_UNIQUE_PACKAGE_NAME
#undef PACKAGE
#define PACKAGE PACKAGE_NEWSCACHECLEAN
#endif

#define MAX(a,b) (((a)>(b))?(a):(b))
#define FILE_ACCESS_TABLE_ENTRIES 100000

Logger slog;
const char *cmnd;
Config Cfg;

class Entry {
public:
	Entry (const char *path, time_t time, long blocks) 
		: Path(path), atime(time), blocks(blocks) {};

	void print (ostream &out) {
		out << Path << " " << atime << " " << blocks << " " << endl;
	};
	
	string Path;
	time_t atime;
	long blocks;
};


vector<Entry*> pvector;
bool sort_entries (const Entry *l, const Entry *r);
void clean(const char *cpath);

struct cache_statistic {
	size_t stPurgedDbCnt;	/* purged .db file count */
	size_t stPurgedDirCnt;	/* purged directory count */
	size_t stPurgedArtCnt;	/* purged .art* file count */
	size_t stPurgedOutArtCnt;
				/* purged outdated .art* file count */
	size_t stDbCnt;		/* overall .db file count before purge */
	size_t stDirCnt;	/* overall directory count before purge */
	size_t stArtCnt;	/* overall .art file count before purge */
	size_t stOutArtCnt;	/* overall outdated .art file count before p.*/
	size_t stPurgedDbSize;	/* purged .db filesize */
	size_t stPurgedArtSize;	/* purged .art* filesize */
	size_t stPurgedOutArtSize;
				/* purged outdated .art* filesize */
	size_t stDbSize;	/* overall .db filesize before purge */
	size_t stArtSize;	/* overall .art* filesize before purge */
	size_t stOutArtSize;	/* overall outdated .art filesize before p. */
};

long _blocks = 0;

static void print_statistic(ostream &out, struct cache_statistic *pS);
void remove_elements(long max_blocks, struct cache_statistic *pS);

static void print_statistic(ostream &out, struct cache_statistic *pS)
{
	out << "PurgedDbCnt: " << pS->stPurgedDbCnt << endl;
	out << "PurgedDirCnt: " << pS->stPurgedDirCnt << endl;
	out << "PurgedArtCnt: " << pS->stPurgedArtCnt << endl;
	out << "PurgedOutArtCnt: " << pS->stPurgedOutArtCnt << endl;
	out << "DbCnt: " << pS->stDbCnt << endl;
	out << "DirCnt: " << pS->stDirCnt << endl;
	out << "ArtCnt: " << pS->stArtCnt << endl;
	out << "OutArtCnt: " << pS->stOutArtCnt << endl;
	out << "PurgedDbSize: " << pS->stPurgedDbSize / 2 /
	    1024 << "Mb" << endl;
	out << "PurgedArtSize: " << pS->stPurgedArtSize / 2 /
	    1024 << "Mb" << endl;
	out << "DbSize: " << pS->stDbSize / 2 / 1024 << "Mb" << endl;
	out << "ArtSize: " << pS->stArtSize / 2 / 1024 << "Mb" << endl;
	out << "OutArtSize: " << pS->stOutArtSize /2 / 1024 << "MB" << endl;
}

bool sort_entries (const Entry *l, const Entry *r)
{
	if (l->atime < r->atime)
		return 1;
	else
		return 0;
}


void remove_elements(long max_blocks, struct cache_statistic *pS)
{
	int isDb;
	char cBuffer[MAXPATHLEN];
	vector<Entry *>::iterator begin, end;

	for (begin=pvector.begin(), end=pvector.end();
	     begin != end; begin++) {
		(*begin)->Path.length();
		if (strcmp((*begin)->Path.c_str() + (*begin)->Path.length() - 3, ".db") == 0) {
			isDb = 1;
			pS->stDbCnt++;
			pS->stDbSize += (*begin)->blocks;
		} else {
			isDb = 0;
			if ((*begin)->atime != 0) {
				pS->stArtCnt++;
				pS->stArtSize += (*begin)->blocks;
			} else {
				pS->stOutArtCnt++;
				pS->stOutArtSize += (*begin)->blocks;
			}
		}
		if ((*begin)->atime != 0 && _blocks <= max_blocks)
			continue;
		if (unlink((*begin)->Path.c_str()) == 0) {
			_blocks -= (*begin)->blocks;
			if (!isDb) {
				if ((*begin)->atime != 0) {
					pS->stPurgedArtCnt++;
					pS->stPurgedArtSize += (*begin)->blocks;
				} else {
					pS->stPurgedOutArtCnt++;
					pS->stPurgedOutArtSize += (*begin)->blocks;
				}
			} else {
				pS->stPurgedDbCnt++;
				pS->stPurgedDbSize += (*begin)->blocks;
				strcpy(cBuffer, (*begin)->Path.c_str());
				cBuffer[strlen(cBuffer) - 3] = '\0';
				if (rmdir(cBuffer) == 0) {
					pS->stPurgedDirCnt++;
				} else if (errno == ENOTEMPTY) {
					slog.p(Logger::Debug) 
						<< "Error rmdir "
						<< "(Directory not empty) "
						<< (*begin)->Path << "\n";
				} else {
					slog.p(Logger::Error) 
						<< "Error rmdir "
						<< (*begin)->Path << "\n";
				}
			}
		} else {
			slog.
			    p(Logger::
			      Error) << "Error unlink " << (*begin)->Path << "\n";
		}
	}
}

void clean(const char *cpath)
{
	VERB(slog.p(Logger::Debug) << "clean(" << cpath << ")\n");
	DIR *d;
	time_t dbatime;
	char buf[MAXPATHLEN];

	struct dirent *f;
	struct stat s;

	sprintf(buf, "%s/.db", cpath);
	// Exception Directories
	if (strcmp(buf, "./.artSpool") == 0)
		return;
	if (strcmp(buf, "./.badArticles") == 0)
		return;
	if (stat(buf, &s) == 0) {
		dbatime = MAX(s.st_atime, s.st_mtime);
		pvector.push_back( new Entry (buf, dbatime, s.st_blocks) );
		_blocks += s.st_blocks;
		--dbatime;
	} else {
		dbatime = 0;
	}

	if ((d = opendir(cpath)) == NULL) {
		slog.p(Logger::Error) << "cannot open " << cpath << "\n";
		return;
	}
	while ((f = readdir(d)) != NULL) {
		if (strcmp(f->d_name, ".") == 0)
			continue;
		if (strcmp(f->d_name, "..") == 0)
			continue;
		sprintf(buf, "%s/%s", cpath, f->d_name);
		if (stat(buf, &s) == 0) {
			if (S_ISDIR(s.st_mode))
				clean(buf);
		} else {
			slog.
			    p(Logger::
			      Error) << "cannot stat " << buf << "\n";
		}
	}
	closedir(d);

	if ((d = opendir(cpath)) == NULL) {
		slog.p(Logger::Error) << "cannot reopen " << cpath << "\n";
		return;
	}
	// sanity check
	string ngname (cpath);
	// ./be/comp/os/unix -> be.comp.os.unix
	ngname.replace (0, 2, "");
	int i;
	while ((i=ngname.find ("/")) != -1) {
		ngname[i] = '.';
	}
	NVNewsgroup *ng;
	OverviewFmt *fmt = new OverviewFmt ();
	int firstnr;
	int lastnr;
	try {
		string dbname = cpath;
		dbname += "/.db";
		struct stat s;
		if (stat (dbname.c_str(), &s) == 0) {
			ng = new NVNewsgroup (fmt,
				(const char *) Cfg.SpoolDirectory,
				ngname.c_str());
		} else {
			ng = NULL;
		}
	} catch (NoSuchGroupError e) {
		ng = NULL;
	}
	// FIXME: what we are doing with bad database files?
	NC_CATCH_ALL ("Error, please report bug!");
	if (ng != NULL) {
		firstnr = ng->firstnbr ();
		lastnr = ng->lastnbr ();
		delete ng;
	}
	string art;
	int nr;
	while ((f = readdir(d)) != NULL) {
		sprintf(buf, "%s/%s", cpath, f->d_name);
		if (stat(buf, &s) == 0) {
			if (S_ISREG(s.st_mode) && dbatime != 0 &&
			    strncmp(f->d_name, ".art", 4) == 0) {
				time_t at = MAX(s.st_atime, s.st_mtime);
				if (dbatime <= at)
					at = dbatime;
				art = f->d_name;
				art.replace (0,4, "");
				if (sscanf (art.c_str(), "%d", &nr) != 1) {
					slog.p(Logger::Error) << "error "
						<< "parsing " << f->d_name;
				} else if ( ng == NULL | nr < firstnr || nr > lastnr) {
					at = 0;
				}
				pvector.push_back (new Entry(buf, at, 
					s.st_blocks) );
				_blocks += s.st_blocks;
			}
		} else {
			slog.
			    p(Logger::
			      Error) << "cannot stat " << buf << "\n";
		}
	}
	closedir(d);
	return;
}



#define USAGE \
"       -h, --help\n"\
"              Show summary of options.\n"\
"\n"\
"       -v, --version\n"\
"              Show version of program.\n"\
"\n"\
"       -c --configuration config-file\n"\
"\n"\
"       -s --statistic\n"\
"              Show statistic informations.\n"\
"\n"\
"       -p --print-purgetable\n"\
"              Print the sorted purge table.\n"\
"\n"\
"       -t --try\n"\
"              Try, not really removing files.\n"\


int main(int argc, char **argv)
{
#ifndef WITH_SYSLOG
	char logfile[MAXPATHLEN];
	time_t t;
	pid_t p;
#endif
	char conffile[MAXPATHLEN];
	int opt_config = 0;
	int c, aerr = 0;
	struct cache_statistic Stat;
	int nice;
	int stat=0;
	int try_flag = 0;
	char p_flag=0;
	vector<Entry *>::iterator begin, end;

	memset ((void *) &Stat, 0, sizeof (struct cache_statistic));

	sprintf(conffile, "%s/newscache.conf", SYSCONFDIR);

	cmnd = argv[0];
	while (1) {
		int option_index = 0;
		static struct option long_options[] = {
			{"version", 0, 0, 'v'},
			{"help", 0, 0, 'h'},
			{"configuration", 1, 0, 'c'},
			{"statistic", 0, 0, 's'},
			{"try", 0, 0, 't'},
			{"print-purgetable", 0, 0, 'p'},
			{0, 0, 0, 0}
		};

		c = getopt_long (argc, argv, "vhc:stp", long_options,
				&option_index);

		if (c == -1)
			break;

		switch (c) {
			case 'v':
				cout << PACKAGE << " " << VERSION << endl;
				exit (0);
				break;

			case 'h':
				cout << "Usage: " << cmnd << endl;
				cout << USAGE << endl;
				exit (0);
				break;

			case 'c':
				strcpy (conffile, optarg);
				++opt_config;
				break;

			case 's':
				stat = 1;
				break;

			case 't':
				try_flag = 1;
				break;
			
			case 'p':
				p_flag = 1;
				break;

			default:
				aerr = 1;
				break;
		}
	}

	if (aerr || optind != argc) {
		cerr << "Usage: " << cmnd << " [options]\n" << USAGE;
		exit(1);
	}

	try {
		Cfg.read(conffile);
//    strcpy(nntp_hostname,Cfg.Hostname);
	}
	catch(IOError & io) {
		cerr << "unexpected EOF in " << conffile << "\n";
		exit(2);
	}
	catch(SyntaxError & se) {
		cerr << se._errtext << "\n";
		exit(2);
	}

#ifdef WITH_SYSLOG
	slog.open(PACKAGE, LOG_NDELAY | LOG_PID, LOG_NEWS);
#else
	sprintf (logfile, "%s/newscache_newscacheclean.log", Cfg.LogDirectory);
	slog.open (logfile);
#endif
	slog.p(Logger::Info) << "cleaning spool directory\n";


	/* set nice value for master server */
#ifdef HAVE_SETPRIORITY
	errno = 0;
	nice = Cfg.NiceServer;
	nice += getpriority(PRIO_PROCESS, 0);
	if (nice == -1 && errno != 0) {
		slog.
		    p(Logger::
		      Error) << "getpriority failed: " << strerror(errno)
		    << "\n";
	} else {
		if (setpriority(PRIO_PROCESS, 0, nice) == -1)
			slog.
			    p(Logger::
			      Error) << "setpriority failed: " <<
			    strerror(errno) << "\n";
	}
#endif

	if (chdir(Cfg.SpoolDirectory) < 0) {
		slog.
		    p(Logger::
		      Error) <<
		    "cannot change to spool directory -- exiting\n";
		exit(1);
	}
	if (getuid() == CONF_UIDROOT)
		setugid(Cfg.Username, Cfg.Groupname);
	clean(".");
	sort (pvector.begin(), pvector.end(), sort_entries);
	if (p_flag) {
		cout << "---------------- Purge table ----------------" << endl;
		for (begin=pvector.begin(), end=pvector.end();
		     begin != end; begin++) {
			(*begin)->print (cout);
		}
		cout << "---------------- Purge table ----------------" << endl;
	}
	if (!try_flag) {
		remove_elements(Cfg.SpoolSize * 2, &Stat);
	}
	if (stat) {
		print_statistic(cout, &Stat);
	}

	slog.p(Logger::Info) << "|SpoolDirectory|="
	    << (unsigned int) (_blocks / 2) << "K\n";
	return 0;
}
